core/src/consensus.rs

```diff
@@ -864,16 +879,39 @@ impl Tower {
         );
 
         if let Some(last_voted_slot) = self.last_voted_slot() {
-            if slot_history.check(last_voted_slot) == Check::TooOld {
-                // We could try hard to anchor with other older votes, but opt to simplify the
-                // following logic
-                return Err(TowerError::TooOldTower(
+            if tower_root <= replayed_root {
+                // Normally, we goes into this clause with possible help of reconcile_blockstore_roots_with_tower()
+                if slot_history.check(last_voted_slot) == Check::TooOld {
+                    // We could try hard to anchor with other older votes, but opt to simplify the
+                    // following logic
+                    return Err(TowerError::TooOldTower(
+                        last_voted_slot,
+                        slot_history.oldest(),
+                    ));
+                }
+                self.adjust_lockouts_with_slot_history(slot_history)?;
+                self.initialize_root(replayed_root);
+            } else {
+                let message = format!(
+                    "For some reason, we're REPROCESSING already voted and ROOTED slots; \
+                     VOTING will be SUSPENDED UNTIL {}!",
                     last_voted_slot,
-                    slot_history.oldest(),
-                ));
+                );
+                error!("{}", message);
+                datapoint_error!("tower_error", ("error", message, String));
```

Comment:
Originally, I was planning to abort the `solana-validator`. But I think we should prefer operation easiness over perfection of uptime. I'll write about this a bit later.

Anyway, might be good adding `solana-validator --force-reboot`?

Reply:
this is original implementation: https://github.com/ryoqun/solana/commit/d8c838686fb0493320977a0889cafcacb0e973a3#r43557231

Reply:
> Originally, I was planning to abort the `solana-validator`. But I think we should prefer operation easiness over perfection of uptime. I'll write about this a bit later.

@carllin Generally speaking, I'm at the rigorous school of fail-first-loudly. So, It's usual for me to abort here because future persistent tower is really odd thing at the very least. 

However, I bent my usual philosophy this time. Because correcting this manually is rather hard (for rough idea of recommended recovery steps, see: https://github.com/ryoqun/solana/commit/d8c838686fb0493320977a0889cafcacb0e973a3#r43557231). And I think we shouldn't require/expect validator operators to have that level of skill and acquittance of our validator workings. Instead, I think they should care about their external environment like DC, security, operation assertiveness, etc.

Based on this assumption, I thought it's worthwhile to make validator robust and self-healing as much as possible with being low-maintainance as long as it doesn't increase the likelihood of slashing. Fortunately, we can make this case so, just by allowing future tower. Thus, here is the reason of current form of this pr.


I think usual occurrences of future tower will be caused by unclean machine shutdown or kill -9 or no free space. So, I guess the gap of slots between blockstore and tower will relatively be  small. On the other hand, I think mixed tower <=> blockstore due to manual mis operation should be very rare, compared to the crashes.

So, I think the opportunity cost (due to required human introspection) would outweigh the right and good practice of correcting and validating the aborted validator. Doing the right and correct here doesn't make sense economically if they can just recover just by waiting a min or several.

Then, the only remaining concern of not aborting is there are false negatives for the detection of erroneous conditions.

But, I think this is Ok. Even if validator don't immediately abort on future tower, I think we can expect operators can detect significant downtime of their validator with `solana-watchtower` and other their methods. Because largely-gaped future tower disables voting for significant duration.

And here is the unavoidable false negative: validator could indefinitely be stalled if an operator mixes up saved tower file and the saved tower is from future and they use save identity key across different clusters and they don't use `solana-watchtower`. I think this is so rare to occur in the wild. And we could safely blame operators if that should occur. ;)

Reply:
I think another alternative to this is if we signal the operator that they must go fetch a snapshot > their tower root, it might be easier to reason about in the future as this grows in complexity 👻

Reply:
> I think another alternative to this is if we signal the operator that they must go fetch a snapshot >= their tower root, it might be easier to reason about in the future as this grows in complexity

You're right. I also considered it.

https://github.com/ryoqun/solana/commit/d8c838686fb0493320977a0889cafcacb0e973a3#diff-8aa35f64ce162f53fcdda9823deb9a87ac8a0c35d0738533e35c2241c726e769R755-R756:

>                          use snapshot (newer than {}) fetched from the cluster by running without \
>                         --no-snapshot-fertch or populate missing data in blockstore by running \


But, recommending newer snapshot might not be an option for seriously operating _mainnet_ validators. That's because they have to trust the newer snapshot once more after their initial TOFU. (that's because secured snapshot isn't still a thing) And I hope they insist to keep their ledger integrity so even after the catastrophic failure like this, as the guardian of quite large amount of assets..

So, I took the red pill (complexity at our side). ;) I think these still are manageable.